src/watch_state.rs

```diff
@@ -0,0 +1,231 @@
+use std::sync::{Arc, Mutex};
+
+use watchexec::action::Outcome;
+
+/// Tracks the current state of spin watch. Handles transitioning states and converting effects
+/// into outcomes based on the current state.
+#[derive(Debug, Clone)]
+pub struct StateMachine {
+    state: Arc<Mutex<State>>,
+    clear: bool,
+    skip_build: bool,
+}
+
+impl StateMachine {
+    pub fn new(skip_build: bool, clear: bool) -> Self {
+        Self {
+            state: Arc::new(Mutex::new(match skip_build {
+                false => State::Building,
+                true => State::Running,
+            })),
+            clear,
+            skip_build,
+        }
+    }
+
+    /// Get the current state of the state machine.
+    pub fn get_state(&self) -> State {
+        *self.state.lock().unwrap()
+    }
+
+    /// Based on the given effect return the correct outcome. Transition the internal state of the
+    /// machine if necessary.
+    pub fn handle(&self, effect: Effect) -> Outcome {
+        let mut state = self.state.lock().unwrap();
+        tracing::debug!("handling effect {:?} in current state {:?}", effect, *state);
+        let outcome = match (effect, *state) {
+            (Effect::Exit, _) => {
+                Outcome::if_running(Outcome::both(Outcome::Stop, Outcome::Exit), Outcome::Exit)
+            }
+            (Effect::ProcessFailed, _) => Outcome::if_running(Outcome::Stop, Outcome::DoNothing),
+            (Effect::ProcessCompleted, State::Building) => {
+                *state = State::Running;
+                self.restart_outcome(true)
+            }
+            (Effect::ProcessCompleted, State::Running) => Outcome::DoNothing,
+            (Effect::ManifestChange, State::Building) => self.restart_outcome(false),
+            (Effect::ManifestChange, State::Running) => {
+                if !self.skip_build {
+                    *state = State::Building;
+                }
+                self.restart_outcome(false)
+            }
+            (Effect::SourceChange, State::Building) => self.restart_outcome(false),
+            (Effect::SourceChange, State::Running) => {
+                if !self.skip_build {
+                    *state = State::Building;
+                    return self.restart_outcome(false);
+                }
+                Outcome::DoNothing
+            }
+            (Effect::ArtifactChange, State::Building) => Outcome::DoNothing,
+            (Effect::ArtifactChange, State::Running) => self.restart_outcome(false),
+            (Effect::DoNothing, _) => Outcome::if_running(Outcome::DoNothing, Outcome::Start),
+        };
+        tracing::debug!("now in {:?} state with outcome {:?}", *state, outcome);
+        outcome
+    }
+
+    fn restart_outcome(&self, prevent_clear: bool) -> Outcome {
```

Comment:
The boolean here is not clear at the call site (I was reading the state machine going "what are all these false values"). It may seem a bit laborious but what about an enum a la `Clear::Allow` / `Clear::Prevent`?

Reply:
I was debating if this deserved an enum and I decided it didn't. On second thought I agree that it does.

Reply:
I actually opted for consts. https://stackoverflow.com/questions/52545734/is-there-a-way-to-use-a-boolean-as-an-underlying-type-for-an-enum.

I felt mapping between bool and an enum was a little clunky.